### ISSUE TITLE
Markdown support in beatmap discussions

### ISSUE THREAD

**Issue by peppy** [ppy/osu-web#451]
Mod comments should be able to use markdown. Possibly with headings disallowed.

**Comment by Ephemeralis** [comment]
I'm told that markdown support for moddingv2 is currently in, but not enabled?

![](https://cdn.discordapp.com/attachments/275879118799896577/538155860678148118/unknown.png)

The full ensemble doesn't need to be in, but the basics like bold, italics, underline, strikethrough, external links, bullet and numbered lists would be a huge bonus to usability for those used to the syntax.

**Comment by dorothy3242** [comment]
From what I can tell, markdown suppport is also present in reviews. Having it in part of the modding panel features and not the other seems inconsistent and makes this all the weirder. Being able to embed images which would display small but could be clicked to display large size in a gallery format (how it's handled for forum posts/ beatmap descriptions right now) would be especially useful for modding instead of always needing to click a link opening a different page/website to view visual suggestions. It feels like this is really a key feature that we've learned to do with out but is still sorely missing.

Related issue https://github.com/ppy/osu-web/issues/6082 also discusses this a bit

**Comment by dorothy3242** [comment]
I'd like to get an update on this issue if possible. It's really wrong it's only partially present, by being in reviews and in beatmap comments, yet is still missing from modding posts themselves after nearly 5 years when it would be so useful.

**Comment by peppy** [comment]
Going to add this to the next milestone. We should definitely get the basic style codes in at very least, if not full support (minus headings as suggested).

**Comment by nanaya** [comment]
this will be fun in term of not killing the browser rendering hundreds of markdown

**Comment by peppy** [comment]
Is it the parsing step that is going to add overhead? Wondering if we just hack in a very simple implementation to get just the basic formatting working (italic / bold).

We can't just render on-screen elements because the post-render height is being used for the overall page layout, right?

**Comment by dorothy3242** [comment]
I'm not technically well versed but how is it it works for comments but would have rendering issues for modding posts? Is there a key difference between them causing pain points?

**Comment by nanaya** [comment]
Comments are paginated but modding discussions may have hundreds of posts loaded at once.

**Comment by dorothy3242** [comment]
Would paginating modding posts in a similar method as how beatmaps in search are loaded be possible?

**Comment by VINXIS** [comment]
auto refresh/infinite scroll as u go up/down the different sections in the modding page would be good for the modding discussion page no? Probably related to https://github.com/ppy/osu-web/issues/6285

**Comment by peppy** [comment]
As soon as you start using placeholders, it means you can't use browser search to search the page for text. This is one of the main reasons we haven't considered this until now. Would that not be considered an issue for modding?

**Comment by dorothy3242** [comment]
With the modding and status filters available as well as visually seeing the timeline for each diff, browser search is not really needed. The main issue I can imagine is how it would relate to long discussions where 1 post has many many many replies. That is a case where ctrl+f is useful and I'm not sure how those posts would be loaded.

**Comment by peppy** [comment]
I guarantee even if you are happy with the functionality being removed, there will be people that depend on it.

**Comment by dorothy3242** [comment]
In that case would adding functionality to search map discussions the same way you can search specific forum threads be an idea?

**Comment by peppy** [comment]
Yes, that is possible. It's also a much larger implementation, involving potentially new infrastructure to handle it.

**Comment by dorothy3242** [comment]
I see. Is there somewhere to view future design changes and how they're intended to work to get a better idea of the plan?

Having markdown especially for allowing image embeds is very, very important. If that could at least be added to reviews it would be a good step forward and reviews have the basic markdown support so far already.

Reviews already have bold/italic for the paragraph parts, just posts don't.

However I'm not sure putting it the main thing for markdown on posts behind other steps like more search etc. is the best choice either when they're not uh, directly related. Which would leave it to that tradeoff of if people value markdown or ctrl+f more, from what I understand. Would a survey on that be helpful?

**Comment by peppy** [comment]
@nanaya as per https://github.com/ppy/osu-web/issues/451#issuecomment-801231558, what's the difficulty level of getting markdown support only in the reviews section? i know that uses its own editor and database structure, so wondering if it's a starting point to get this in.

**Comment by nanaya** [comment]
it's already markdown internally but because the editor itself is a visual editor, it needs ui for actually styling or adding things to them (among other things)

**Comment by peppy** [comment]
> Is there anything in particular holding this back for further implementation? Markdown has been sorely needed for a long while now, especially for features like image embedding, etc.

see the posts above. it should explain that nothing is holding this back apart from developer time.

**Comment by Ephemeralis** [comment]
This totally fell by the wayside, but yes, embedding images is *critical* to being able to properly articulate most mods/suggestions/change. Preserving these images indefinitely (for the sake of history) would be good, but is nowhere near important enough to delay having the feature entirely.

The closest analogue I can think of would be having to open GitHub suggestion diffs in an entirely new window inside PRs every time you wanted to see what was in them. Since people have to click image URLs to do this constantly, reviews are an arduous slog at the mercy of the user's tab management.

The NAT have mentioned recently in internal discussions that this continues to be a severe friction point with beatmap discussions use.

**Comment by nanaya** [comment]
In addition to the ui requirement above, there may be additional concern regarding limiting the display of the image (how big can they be, how many are allowed in single post, and are we going to allow collapsing the longer posts and how, among other things). Some of which probably require optimising the page first because otherwise it'll just get even slower.

**Comment by peppy** [comment]
Bumping this to priority zero as it is required to make the system usable, according to the NAT team.

**Comment by peppy** [comment]
@notbakaneko nanaya mentions that you're still working through this, can you give a status update? I think you may need to stop cleanup efforts and just make this happen at this point, since it's taking so long.

Proposals:

- If client side markdown rendering becoming too slow is a concern, do it server side and send the html
- If images pushing pages off-screen is a concern (or initial image content taking too long / too much bandwidth to load), do js level async image loading and remove any seek animations so the amount the page is pushed down is reduced to only stuff on the current screen.

**Comment by notbakaneko** [comment]
It's not as straight forward as just rendering markdown and sending it to the client; there are additional client-side parsers on the discussion page and it turns out these only work in plain text. 
There's already a markdown renderer being used for discussion reviews, except the plugins being used here also only work with the very limited set of markdown being rendered (reviews aren't actually markdown either)

Basically,
- need to update `react-markdown` (out of date, API changes, hard to find documentation for the old format)
- update the related dependencies used by react-markdown that we're also using directly (versions not compatible)
- fix all the breakages from updates
- fix parsing and rendering of our plugins

**Comment by zoey-on-github** [comment]
im sorry if this isn't the right place to say this but why is this tagged both "priority:0" and "low priority"? the tags seem to contradict each oterh

**Comment by notbakaneko** [comment]
markdown support has been added in #9843; requests for additional syntax should open new issues
